build-tools/packages/build-tools/src/type-test-generator/typeTestUtils.ts

```diff
@@ -0,0 +1,260 @@
+ */
+export function getTypeDefinitionFilePath(packageBasePath: string): string | undefined {
+	const previousPackageJsonPath = tryGetPreviousPackageJsonPath(packageBasePath);
+	const packageJson: PackageJson = readJsonSync(previousPackageJsonPath);
+	// Check the exports entries
+	if (packageJson.exports) {
+		return getTypePathFromExport(packageJson, packageBasePath);
+		// Check the types field from the previous package.json as a fallback
+	} else if (packageJson.types) {
+		return path.join(packageBasePath, packageJson.types);
+	} else {
+		throw new Error(
+			`Type definition file path could not be determined from ${previousPackageJsonPath}. No 'exports' nor 'type' fields found.`,
+		);
+	}
+}
+
+/**
+ * Extracts type data from a TS source file and creates a map where each key is a type name and the value is its type data.
+ * @param file - The source code file containing type data
+ * @returns The mapping between item and its type
+ */
+export function typeDataFromFile(file: SourceFile): Map<string, TypeData> {
+	const typeData = new Map<string, TypeData>();
+	const exportedDeclarations = file.getExportedDeclarations();
+
+	for (const declarations of exportedDeclarations.values()) {
+		for (const declaration of declarations) {
+			getNodeTypeData(declaration).forEach((typeDefinition) => {
+				const fullName = getFullTypeName(typeDefinition);
+				if (typeData.has(fullName)) {
+					// This system does not properly handle overloads: instead it only keeps the last signature.
+					console.warn(`skipping overload for ${fullName}`);
+				}
+				typeData.set(fullName, typeDefinition);
+			});
+		}
+	}
+	return typeData;
+}
+
+/**
+ * Initializes TypeScript projects for the current and previous package versions and loads specific source files.
+ * @param typeDefinitionFilePath - The path to the type definition file for the previous version.
+ * @param previousBasePath - The path to the root of the previous package version.
+ * @param previousPackageName - The name of the previous package version.
+ * @returns {{ currentFile: SourceFile, previousFile: SourceFile }} - The loaded source files for the current and previous versions.
+ */
+export function initializeProjectsAndLoadFiles(
+	typeDefinitionFilePath: string,
+	previousBasePath: string,
+	previousPackageName: string,
+) {
+	const currentFile = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: "tsconfig.json",
+	}).getSourceFileOrThrow("index.ts");
+
+	const previousTsConfigPath = path.join(previousBasePath, "tsconfig.json");
+	const project = new Project({
+		skipFileDependencyResolution: true,
+		tsConfigFilePath: existsSync(previousTsConfigPath) ? previousTsConfigPath : undefined,
+	});
+
+	let previousFile: SourceFile;
+	// Check for existence of alpha and add appropriate file
+	if (existsSync(typeDefinitionFilePath)) {
+		project.addSourceFilesAtPaths(typeDefinitionFilePath);
+		previousFile = project.getSourceFileOrThrow(`${previousPackageName}-alpha.d.ts`);
```

Comment:
We should not look at `*alpha.d.ts` or similar paths. Should just work from enumerating `@ff/foo-previous/*` where `*` is any simple path that is not internal as well as `@ff/foo-previous` for the public case.
We abandoned `foo-alpha.d.ts`; so, this is not viable.

Reply:
This should be addressed in my refactor of this PR, https://github.com/microsoft/FluidFramework/pull/20803